// @adobe/pdfservices-node-sdk/lib/pdfjobs/params/exportpdf/ExportPDFTargetFormat.js
"use strict";
/*
 * Copyright 2024 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExportPDFTargetFormat = void 0;
/**
 * Supported target formats for {@link ExportPDFJob}
 */
var ExportPDFTargetFormat;
(function (ExportPDFTargetFormat) {
    /**
     * Represents "application/msword" media type
     */
    ExportPDFTargetFormat["DOC"] = "doc";
    /**
     * Represents "application/vnd.openxmlformats-officedocument.wordprocessingml.document" media type
     */
    ExportPDFTargetFormat["DOCX"] = "docx";
    /**
     * Represents "application/vnd.openxmlformats-officedocument.presentationml.presentation" media type
     */
    ExportPDFTargetFormat["PPTX"] = "pptx";
    /**
     * Represents "text/rtf" media type
     */
    ExportPDFTargetFormat["RTF"] = "rtf";
    /**
     * Represents "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet" media type
     */
    ExportPDFTargetFormat["XLSX"] = "xlsx";
})(ExportPDFTargetFormat || (exports.ExportPDFTargetFormat = ExportPDFTargetFormat = {}));
