// @adobe/pdfservices-node-sdk/lib/pdfjobs/params/exportpdf/ExportOCRLocale.js
"use strict";
/*
 * Copyright 2024 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExportOCRLocale = void 0;
/**
 * Supported locales for {@link ExportPDFJob}
 */
var ExportOCRLocale;
(function (ExportOCRLocale) {
    /**
     * Represents "Danish (Denmark)" locale
     */
    ExportOCRLocale["DA_DK"] = "da-DK";
    /**
     * Represents "Lithuanian (Lithuania)" locale
     */
    ExportOCRLocale["LT_LT"] = "lt-LT";
    /**
     * Represents "Slovenian (Slovenia)" locale
     */
    ExportOCRLocale["SL_SI"] = "sl-SI";
    /**
     * Represents "Greek (Greece)" locale
     */
    ExportOCRLocale["EL_GR"] = "el-GR";
    /**
     * Represents "Russian (Russia)" locale
     */
    ExportOCRLocale["RU_RU"] = "ru-RU";
    /**
     * Represents "English (United States)" locale.
     */
    ExportOCRLocale["EN_US"] = "en-US";
    /**
     * Represents "Hungarian (Hungary)" locale
     */
    ExportOCRLocale["HU_HU"] = "hu-HU";
    /**
     * Represents "Estonian (Estonia)" locale
     */
    ExportOCRLocale["ET_EE"] = "et-EE";
    /**
     * Represents "Portuguese (Brazil)" locale
     */
    ExportOCRLocale["PT_BR"] = "pt-BR";
    /**
     * Represents "Ukrainian (Ukraine)" locale
     */
    ExportOCRLocale["UK_UA"] = "uk-UA";
    /**
     * Represents "Norwegian (Norway)" locale
     */
    ExportOCRLocale["NB_NO"] = "nb-NO";
    /**
     * Represents "Polish (Poland)" locale
     */
    ExportOCRLocale["PL_PL"] = "pl-PL";
    /**
     * Represents "Latvian (Latvia)" locale
     */
    ExportOCRLocale["LV_LV"] = "lv-LV";
    /**
     * Represents "Finnish (Finland)" locale
     */
    ExportOCRLocale["FI_FI"] = "fi-FI";
    /**
     * Represents "Japanese (Japan)" locale
     */
    ExportOCRLocale["JA_JP"] = "ja-JP";
    /**
     * Represents "Spanish (Spain)" locale
     */
    ExportOCRLocale["ES_ES"] = "es-ES";
    /**
     * Represents "Bulgarian (Bulgaria)" locale
     */
    ExportOCRLocale["BG_BG"] = "bg-BG";
    /**
     * Represents "English (United Kingdom)" locale
     */
    ExportOCRLocale["EN_GB"] = "en-GB";
    /**
     * Represents "Czech (Czech Republic)" locale
     */
    ExportOCRLocale["CS_CZ"] = "cs-CZ";
    /**
     * Represents "Maltese (Malta)" locale
     */
    ExportOCRLocale["MT_MT"] = "mt-MT";
    /**
     * Represents "German (Germany)" locale
     */
    ExportOCRLocale["DE_DE"] = "de-DE";
    /**
     * Represents "Croatian (Croatia)" locale
     */
    ExportOCRLocale["HR_HR"] = "hr-HR";
    /**
     * Represents "Slovak (Slovakia)" locale
     */
    ExportOCRLocale["SK_SK"] = "sk-SK";
    /**
     * Represents "Serbian (Serbia)" locale
     */
    ExportOCRLocale["SR_SR"] = "sr-SR";
    /**
     * Represents "Catalan (Canada)" locale
     */
    ExportOCRLocale["CA_CA"] = "ca-CA";
    /**
     * Represents "Macedonian (Macedonia)" locale
     */
    ExportOCRLocale["MK_MK"] = "mk-MK";
    /**
     * Represents "Korean (Korea)" locale
     */
    ExportOCRLocale["KO_KR"] = "ko-KR";
    /**
     * Represents "German (Switzerland)" locale
     */
    ExportOCRLocale["DE_CH"] = "de-CH";
    /**
     * Represents "Dutch (Netherlands)" locale
     */
    ExportOCRLocale["NL_NL"] = "nl-NL";
    /**
     * Represents "Chinese (China)" locale
     */
    ExportOCRLocale["ZH_CN"] = "zh-CN";
    /**
     * Represents "Swedish (Sweden)" locale
     */
    ExportOCRLocale["SV_SE"] = "sv-SE";
    /**
     * Represents "Italian (Italy)" locale
     */
    ExportOCRLocale["IT_IT"] = "it-IT";
    /**
     * Represents "Turkish (Turkey)" locale
     */
    ExportOCRLocale["TR_TR"] = "tr-TR";
    /**
     * Represents "French (France)" locale
     */
    ExportOCRLocale["FR_FR"] = "fr-FR";
    /**
     * Represents "Romanian (Romania)" locale
     */
    ExportOCRLocale["RO_RO"] = "ro-RO";
    /**
     * Represents "Hebrew (Israel)" locale
     */
    ExportOCRLocale["IW_IL"] = "iw-IL";
    /**
     * Represents "European Portuguese" locale
     */
    ExportOCRLocale["PT_PT"] = "pt-PT";
    /**
     * Represents "Norwegian Nynorsk (Norway)" locale
     */
    ExportOCRLocale["NN_NO"] = "nn-NO";
    /**
     * Represents "Chinese (traditional)" locale
     */
    ExportOCRLocale["ZH_HANT"] = "zh-Hant";
    /**
     * Represents "Basque (Spain)" locale
     */
    ExportOCRLocale["EU_ES"] = "eu-ES";
    /**
     * Represents "Galician (Spain)" locale
     */
    ExportOCRLocale["GL_ES"] = "gl-ES";
})(ExportOCRLocale || (exports.ExportOCRLocale = ExportOCRLocale = {}));
