// @adobe/pdfservices-node-sdk/lib/internal/dto/response/PDFServicesApiAutotagResponse.js
"use strict";
/*
 * Copyright 2024 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PDFServicesApiAutotagResponse = void 0;
const CloudAsset_1 = require("../../../io/CloudAsset");
const ObjectUtil_1 = require("../../util/ObjectUtil");
class PDFServicesApiAutotagResponse {
    constructor(taggedPDF, report, resource, errorResponse, status) {
        this._taggedPDF = taggedPDF;
        this._report = report;
        this._resource = resource;
        this._status = status;
        this._errorResponse = errorResponse;
    }
    get taggedPDF() {
        if (!ObjectUtil_1.ObjectUtil.isUndefinedOrNull(this._taggedPDF))
            return new CloudAsset_1.CloudAsset(this._taggedPDF.assetID, this._taggedPDF.downloadUri);
    }
    get report() {
        if (!ObjectUtil_1.ObjectUtil.isUndefinedOrNull(this._report))
            return new CloudAsset_1.CloudAsset(this._report.assetID, this._report.downloadUri);
    }
    get resource() {
        if (!ObjectUtil_1.ObjectUtil.isUndefinedOrNull(this._resource)) {
            return new CloudAsset_1.CloudAsset(this._resource.assetID, this._resource.downloadUri);
        }
    }
    get status() {
        return this._status;
    }
    get errorResponse() {
        return this._errorResponse;
    }
}
exports.PDFServicesApiAutotagResponse = PDFServicesApiAutotagResponse;
