// @adobe/pdfservices-node-sdk/lib/pdfjobs/result/pdfproperties/Content.js
"use strict";
/*
 * Copyright 2024 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Content = void 0;
/**
 * This class provides information about the content of the specified PDF file, such as number of images, textual
 * content etc.
 */
class Content {
    /**
     * Returns the number of images in the content.
     *
     * @returns An Integer denoting the number of images in the content
     */
    get numberOfImages() {
        return this._numberOfImages;
    }
    /**
     * Returns a boolean specifying whether the content consists only images.
     *
     * @returns `true` if the content consists only images. False otherwise.
     */
    get containsImagesOnly() {
        return this._onlyImages;
    }
    /**
     * Returns a boolean specifying whether the content has any text.
     *
     * @returns `true` if the content has any text. False otherwise.
     */
    get containsText() {
        return this._hasText;
    }
    /**
     * Returns a boolean specifying whether the content has any images.
     *
     * @returns `true` if the content has any images. False otherwise.
     */
    get containsImages() {
        return this._hasImages;
    }
    /**
     * Returns a boolean specifying whether the content is empty/blank.
     *
     * @returns `true` if the content is empty. False otherwise.
     */
    get isEmpty() {
        return this._isEmpty;
    }
}
exports.Content = Content;
