// @adobe/pdfservices-node-sdk/lib/pdfjobs/params/createpdf/DocumentLanguage.js
"use strict";
/*
 * Copyright 2024 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocumentLanguage = void 0;
/**
 * Supported locales for {@link CreatePDFJob}
 */
var DocumentLanguage;
(function (DocumentLanguage) {
    /**
     * Represents "Danish (Denmark)" locale
     */
    DocumentLanguage["DA_DK"] = "da-DK";
    /**
     * Represents "Lithuanian (Lithuania)" locale
     */
    DocumentLanguage["LT_LT"] = "lt-LT";
    /**
     * Represents "Slovenian (Slovenia)" locale
     */
    DocumentLanguage["SL_SI"] = "sl-SI";
    /**
     * Represents "Greek (Greece)" locale
     */
    DocumentLanguage["EL_GR"] = "el-GR";
    /**
     * Represents "Russian (Russia)" locale
     */
    DocumentLanguage["RU_RU"] = "ru-RU";
    /**
     * Represents "English (United States)" locale.
     */
    DocumentLanguage["EN_US"] = "en-US";
    /**
     * Represents "Chinese (Hong Kong)" locale
     */
    DocumentLanguage["ZH_HK"] = "zh-HK";
    /**
     * Represents "Hungarian (Hungary)" locale
     */
    DocumentLanguage["HU_HU"] = "hu-HU";
    /**
     * Represents "Estonian (Estonia)" locale
     */
    DocumentLanguage["ET_EE"] = "et-EE";
    /**
     * Represents "Portuguese (Brazil)" locale
     */
    DocumentLanguage["PT_BR"] = "pt-BR";
    /**
     * Represents "Ukrainian (Ukraine)" locale
     */
    DocumentLanguage["UK_UA"] = "uk-UA";
    /**
     * Represents "Norwegian (Norway)" locale
     */
    DocumentLanguage["NB_NO"] = "nb-NO";
    /**
     * Represents "Polish (Poland)" locale
     */
    DocumentLanguage["PL_PL"] = "pl-PL";
    /**
     * Represents "Latvian (Latvia)" locale
     */
    DocumentLanguage["LV_LV"] = "lv-LV";
    /**
     * Represents "Finnish (Finland)" locale
     */
    DocumentLanguage["FI_FI"] = "fi-FI";
    /**
     * Represents "Japanese (Japan)" locale
     * @remarks Please note that this locale is only supported for US(default) region.
     */
    DocumentLanguage["JA_JP"] = "ja-JP";
    /**
     * Represents "Spanish (Spain)" locale
     */
    DocumentLanguage["ES_ES"] = "es-ES";
    /**
     * Represents "Bulgarian (Bulgaria)" locale
     */
    DocumentLanguage["BG_BG"] = "bg-BG";
    /**
     * Represents "English (United Kingdom)" locale
     */
    DocumentLanguage["EN_GB"] = "en-GB";
    /**
     * Represents "Czech (Czech Republic)" locale
     */
    DocumentLanguage["CS_CZ"] = "cs-CZ";
    /**
     * Represents "Maltese (Malta)" locale
     */
    DocumentLanguage["MT_MT"] = "mt-MT";
    /**
     * Represents "German (Germany)" locale
     */
    DocumentLanguage["DE_DE"] = "de-DE";
    /**
     * Represents "Croatian (Croatia)" locale
     */
    DocumentLanguage["HR_HR"] = "hr-HR";
    /**
     * Represents "Slovak (Slovakia)" locale
     */
    DocumentLanguage["SK_SK"] = "sk-SK";
    /**
     * Represents "Serbian (Serbia)" locale
     */
    DocumentLanguage["SR_SR"] = "sr-SR";
    /**
     * Represents "Catalan (Canada)" locale
     */
    DocumentLanguage["CA_CA"] = "ca-CA";
    /**
     * Represents "Macedonian (Macedonia)" locale
     */
    DocumentLanguage["MK_MK"] = "mk-MK";
    /**
     * Represents "Korean (Korea)" locale
     */
    DocumentLanguage["KO_KR"] = "ko-KR";
    /**
     * Represents "German (Switzerland)" locale
     */
    DocumentLanguage["DE_CH"] = "de-CH";
    /**
     * Represents "Dutch (Netherlands)" locale
     */
    DocumentLanguage["NL_NL"] = "nl-NL";
    /**
     * Represents "Chinese (China)" locale
     */
    DocumentLanguage["ZH_CN"] = "zh-CN";
    /**
     * Represents "Swedish (Sweden)" locale
     */
    DocumentLanguage["SV_SE"] = "sv-SE";
    /**
     * Represents "Italian (Italy)" locale
     */
    DocumentLanguage["IT_IT"] = "it-IT";
    /**
     * Represents "Norwegian (Norway)" locale
     */
    DocumentLanguage["NO_NO"] = "no-NO";
    /**
     * Represents "Turkish (Turkey)" locale
     */
    DocumentLanguage["TR_TR"] = "tr-TR";
    /**
     * Represents "French (France)" locale
     */
    DocumentLanguage["FR_FR"] = "fr-FR";
    /**
     * Represents "Romanian (Romania)" locale
     */
    DocumentLanguage["RO_RO"] = "ro-RO";
    /**
     * Represents "Hebrew (Israel)" locale
     */
    DocumentLanguage["IW_IL"] = "iw-IL";
})(DocumentLanguage || (exports.DocumentLanguage = DocumentLanguage = {}));
