// @adobe/pdfservices-node-sdk/lib/pdfjobs/params/protectpdf/Permission.js
"use strict";
/*
 * Copyright 2024 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Permission = void 0;
/**
 * Supported document permission types for {@link ProtectPDFJob}
 */
var Permission;
(function (Permission) {
    /**
     * Enables low quality printing of the PDF document
     */
    Permission["PRINT_LOW_QUALITY"] = "PRINT_LOW_QUALITY";
    /**
     * Enables high quality printing of the PDF document
     */
    Permission["PRINT_HIGH_QUALITY"] = "PRINT_HIGH_QUALITY";
    /**
     * Enables all the editing permissions in the PDF document except commenting and page extraction
     */
    Permission["EDIT_CONTENT"] = "EDIT_CONTENT";
    /**
     * Enables insertion, deletion and rotation of pages in a PDF document
     */
    Permission["EDIT_DOCUMENT_ASSEMBLY"] = "EDIT_DOCUMENT_ASSEMBLY";
    /**
     * Enables additions of comments, digital signatures and filling in of forms in a PDF document
     */
    Permission["EDIT_ANNOTATIONS"] = "EDIT_ANNOTATIONS";
    /**
     * Enables filling in of forms, digital signature and creation of template pages in a PDF document
     */
    Permission["EDIT_FILL_AND_SIGN_FORM_FIELDS"] = "EDIT_FILL_AND_SIGN_FORM_FIELDS";
    /**
     * Enables copying of content from the PDF document
     */
    Permission["COPY_CONTENT"] = "COPY_CONTENT";
})(Permission || (exports.Permission = Permission = {}));
