// @adobe/pdfservices-node-sdk/lib/MimeType.js
"use strict";
/*
 * Copyright 2024 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MimeType = void 0;
/**
 * A mapping of Media types and extension.
 * <p>
 * <p>
 * Refer :
 * <a href="https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Complete_list_of_MIME_types">
 *     Mozilla MIME type</a>
 * <a href="https://www.iana.org/assignments/media-types/media-types.xhtml">IANA media types</a>
 */
var MimeType;
(function (MimeType) {
    /**
     * Represents ai mime type.
     */
    MimeType["AI"] = "application/illustrator";
    /**
     * Represents bmp mime type.
     */
    MimeType["BMP"] = "image/bmp";
    /**
     * Represents msword mime type.
     */
    MimeType["DOC"] = "application/msword";
    /**
     * Represents docx mime type.
     */
    MimeType["DOCX"] = "application/vnd.openxmlformats-officedocument.wordprocessingml.document";
    /**
     * Represents gif mime type.
     */
    MimeType["GIF"] = "image/gif";
    /**
     * Represents html mime type.
     */
    MimeType["HTML"] = "text/html";
    /**
     * Represents indd mime type.
     */
    MimeType["INDD"] = "application/x-indesign";
    /**
     * Represents jpeg mime type.
     */
    MimeType["JPEG"] = "image/jpeg";
    /**
     * Represents jpg mime type.
     */
    // eslint-disable-next-line @typescript-eslint/no-duplicate-enum-values
    MimeType["JPG"] = "image/jpeg";
    /**
     * Represents pdf mime type.
     */
    MimeType["PDF"] = "application/pdf";
    /**
     * Represents png mime type.
     */
    MimeType["PNG"] = "image/png";
    /**
     * Represents ppt mime type.
     */
    MimeType["PPT"] = "application/vnd.ms-powerpoint";
    /**
     * Represents pptx mime type.
     */
    MimeType["PPTX"] = "application/vnd.openxmlformats-officedocument.presentationml.presentation";
    /**
     * Represents psd mime type.
     */
    MimeType["PSD"] = "image/vnd.adobe.photoshop";
    /**
     * Represents rtf mime type.
     */
    MimeType["RTF"] = "text/rtf";
    /**
     * Represents tif mime type.
     */
    MimeType["TIF"] = "image/tiff";
    /**
     * Represents tiff mime type.
     */
    // eslint-disable-next-line @typescript-eslint/no-duplicate-enum-values
    MimeType["TIFF"] = "image/tiff";
    /**
     * Represents txt mime type.
     */
    MimeType["TXT"] = "text/plain";
    /**
     * Represents xls mime type.
     */
    MimeType["XLS"] = "application/vnd.ms-excel";
    /**
     * Represents xlsx mime type.
     */
    MimeType["XLSX"] = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet";
    /**
     * Represents zip mime type.
     */
    MimeType["ZIP"] = "application/zip";
    /**
     * Represents json mime type.
     */
    MimeType["JSON"] = "application/json";
    /**
     * Represents csv mime type.
     */
    MimeType["CSV"] = "text/csv";
    /**
     * Represents svg mime type.
     */
    MimeType["SVG"] = "image/svg+xml";
})(MimeType || (exports.MimeType = MimeType = {}));
