// @adobe/pdfservices-node-sdk/lib/pdfjobs/params/electronicseal/AppearanceItem.js
"use strict";
/*
 * Copyright 2024 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppearanceItem = void 0;
/**
 * Supported elements to represent electronic seal required for {@link AppearanceOptions}.
 */
var AppearanceItem;
(function (AppearanceItem) {
    /**
     * Represents the name of the certificate owner.
     */
    AppearanceItem["NAME"] = "NAME";
    /**
     * Represents the date of applying the electronic seal.
     */
    AppearanceItem["DATE"] = "DATE";
    /**
     * Represents the distinguished name information of the certificate.
     */
    AppearanceItem["DISTINGUISHED_NAME"] = "DISTINGUISHED_NAME";
    /**
     * Represents labels for seal information.
     */
    AppearanceItem["LABELS"] = "LABELS";
    /**
     * Represents the background image to be used for sealing.
     */
    AppearanceItem["SEAL_IMAGE"] = "SEAL_IMAGE";
})(AppearanceItem || (exports.AppearanceItem = AppearanceItem = {}));
