// @adobe/pdfservices-node-sdk/lib/internal/dto/response/pdfproperties/EncryptionSettingsProperties.js
"use strict";
/*
 * Copyright 2024 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EncryptionSettingsProperties = void 0;
class EncryptionSettingsProperties {
    constructor(encryption) {
        this._algorithm = encryption?.algorithm;
        this._bitLength = encryption?.bit_length;
        this._encryptAttachmentsOnly = encryption?.encrypt_attachments_only;
        this._encryptMetadata = encryption?.encrypt_metadata;
        this._hasOwnerPassword = encryption?.has_owner_password;
        this._hasUserPassword = encryption?.has_user_password;
    }
    get onlyAttachmentsEncrypted() {
        return this._encryptAttachmentsOnly;
    }
    get isOwnerPasswordSet() {
        return this._hasOwnerPassword;
    }
    get isMetadataEncrypted() {
        return this._encryptMetadata;
    }
    get isUserPasswordSet() {
        return this._hasUserPassword;
    }
    get bitLength() {
        return this._bitLength;
    }
    get algorithm() {
        return this._algorithm;
    }
}
exports.EncryptionSettingsProperties = EncryptionSettingsProperties;
