// @adobe/pdfservices-node-sdk/lib/internal/dto/response/pdfproperties/InfoDictProperties.js
"use strict";
/*
 * Copyright 2024 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.InfoDictProperties = void 0;
class InfoDictProperties {
    constructor(info_dict) {
        this._creationDate = info_dict?.CreationDate;
        this._creator = info_dict?.Creator;
        this._modDate = info_dict?.ModDate;
        this._producer = info_dict?.Producer;
        this._author = info_dict?.Author;
        this._title = info_dict?.Title;
    }
    get creationDate() {
        return this._creationDate;
    }
    get creator() {
        return this._creator;
    }
    get modDate() {
        return this._modDate;
    }
    get producer() {
        return this._producer;
    }
    get author() {
        return this._author;
    }
    get title() {
        return this._title;
    }
}
exports.InfoDictProperties = InfoDictProperties;
