// @adobe/pdfservices-node-sdk/lib/internal/dto/response/pdfproperties/PageProperties.js
"use strict";
/*
 * Copyright 2024 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PageProperties = void 0;
const ContentProperties_1 = require("./ContentProperties");
class PageProperties {
    constructor(page) {
        this._pageNumber = page?.page_number;
        this._isScanned = page?.is_scanned;
        this._hasStructure = page?.has_structure;
        this._height = page?.height;
        this._width = page?.width;
        this._content = page?.content ? new ContentProperties_1.ContentProperties(page.content) : undefined;
    }
    get pageNumber() {
        return this._pageNumber;
    }
    get isScanned() {
        return this._isScanned;
    }
    get width() {
        return this._width;
    }
    get hasStructure() {
        return this._hasStructure;
    }
    get content() {
        return this._content;
    }
    get height() {
        return this._height;
    }
}
exports.PageProperties = PageProperties;
