// @adobe/pdfservices-node-sdk/lib/internal/http/HttpBaseResponse.js
"use strict";
/*
 * Copyright 2024 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpBaseResponse = void 0;
class HttpBaseResponse {
    constructor(params) {
        this._statusCode = params.statusCode;
        this._headers = params.headers;
        this._baseResponseDto = params.baseResponseDto;
        this._streamList = params.streamList;
        this._responseContent = params.responseContent;
    }
    get statusCode() {
        return this._statusCode;
    }
    get headers() {
        return this._headers;
    }
    get baseResponseDto() {
        return this._baseResponseDto;
    }
    get streamList() {
        return this._streamList;
    }
    get responseContent() {
        return this._responseContent;
    }
}
exports.HttpBaseResponse = HttpBaseResponse;
