// @adobe/pdfservices-node-sdk/lib/internal/Logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const log4js_1 = __importDefault(require("log4js"));
const fs_1 = __importDefault(require("fs"));
const GlobalConfig_1 = require("./GlobalConfig");
const defaultLogConfig = {
    appenders: {
        console: {
            type: "console",
            layout: {
                type: "pattern",
                pattern: "%d:[%p]: %m"
            }
        }
    },
    categories: {
        default: {
            appenders: ["console"],
            level: "info"
        }
    }
}, initializeLogger = () => {
    if (fs_1.default.existsSync(GlobalConfig_1.GlobalConfig.LOG_FILE_NAME)) {
        console.log(`Logging configuration found at ${GlobalConfig_1.GlobalConfig.LOG_FILE_NAME}.`);
        log4js_1.default.configure(GlobalConfig_1.GlobalConfig.LOG_FILE_NAME);
    }
    else {
        console.log("No logging configuration. Using default config");
        log4js_1.default.configure(defaultLogConfig);
    }
    return log4js_1.default.getLogger();
};
exports.default = initializeLogger();
